lib/github.dart

```
@@ -0,0 +1,1029 @@
+      window.localStorage['github_oauth_token'] ?? '';
+
+  set avatarUrl(String url) {
+    if (url.isNotEmpty) {
+      window.localStorage['github_avatar_url'] = url;
+    } else {
+      window.localStorage.remove('github_avatar_url');
+    }
+  }
+
+  String get avatarUrl => window.localStorage['github_avatar_url'] ?? '';
+
+  set userLogin(String login) {
+    if (login.isNotEmpty) {
+      window.localStorage['github_user_login'] = login;
+    } else {
+      window.localStorage.remove('github_user_login');
+    }
+  }
+
+  String get userLogin => window.localStorage['github_user_login'] ?? '';
+
+  static const _chars =
+      'AaBbCcDdEeFfGgHhIiJjKkLlMmNnOoPpQqRrSsTtUuVvWwXxYyZz1234567890';
+  final Random _rnd = Random.secure();
+
+  String getRandomString(int length) => String.fromCharCodes(Iterable.generate(
+      length, (_) => _chars.codeUnitAt(_rnd.nextInt(_chars.length))));
+
+  String makeRandomSecureAuthInitiationUrl(String baseUrl) {
+    // Create random state string which will be used by GH OAuth and then by
+    // us to encrypt returned gh auth token.
+    final String state = getRandomString(40);
+
+    // Store state in localStorage, because we are going to need it to Decrypt
+    // the returned authorization token.
+    window.localStorage['github_random_state'] = state;
+
+    if (baseUrl.endsWith('/')) {
+      return '$baseUrl$state';
+    } else {
+      return '$baseUrl/$state';
+    }
+  }
+
+  String decryptAuthTokenFromReturnedSecureAuthToken(
+      String encryptedBase64AuthToken) {
+    // Retrieve the random state string we made for the original request in
+    // makeRandomSecureAuthInitiationUrl().  Our auth token was encrypted using
+    // this before sending it back to us, so use it to decrypt.
+    final String randomStateWeSent =
+        window.localStorage['github_random_state'] ?? '';
+
+    try {
+      if (randomStateWeSent.isEmpty) {
+        return 'ERROR-no stored initial state';
```

Should we throw an exception instead?

---

I have changed this to throw exceptions, as these would be highly unusual events.